### ISSUE TITLE
pnpm support

### ISSUE THREAD

**Issue by aleclarson** [lerna/lerna#1818]
`pnpm` support was briefly discussed here: https://github.com/lerna/lerna/issues/1349#issuecomment-384058683

After searching the issues/PRs, it seems there's been no progress yet, so I figured it needs its own issue to track its development and gauge demand.

/cc @zkochan @evocateur

**Comment by evocateur** [comment]
Thanks for opening this, you are correct: there has been no progress. I haven't had the time to devote to investigation, personally. Contributions welcome!

**Comment by evocateur** [comment]
I don't know why stale bot failed to read its own config. This isn't stale.

**Comment by evocateur** [comment]
This is your last warning, stale bot.

**Comment by fineline** [comment]
I'd be really interested in this too. I'd like to use lerna, but I'm not going back to npm / yarn.

**Comment by zkochan** [comment]
@fineline maybe you could try to contribute it to lerna? I will help if needed but I don't have the capacity at the moment to work on this.

**Comment by fineline** [comment]
@zkochan After further research I found that using `pnpm recursive` gives me everything I was looking for in monorepo support, so for me there is no more need for lerna. Thanks.

**Comment by evocateur** [comment]
I experimented with converting lerna/lerna to pnpm yesterday, went pretty smoothly all things considered: https://github.com/lerna/lerna/compare/pnpm-noodling

@zkochan I love the recent progress on pnpm, very exciting! I had one question, how do you make the releases in the pnpm monorepo? I wasn't able to find any scripts or other guidance in the docs, but the diffs of the publish commits look _very_ similar to Lerna output (modulo obvious pnpm differences).

My pnpm noodling branch still needs some work to make `lerna` understand `pnpm recursive` patterns, but it looks 80% of the way there already, from my brief (possibly incorrect) investigation. Among the problems I had was bad dependencies with implicit requires (dependent on npm 3+ flattening to resolve), so any pointers to suitable replacements for `npm-registry-fetch` and `libnpmaccess` would be appreciated. (otherwise i'll probably just low-key fork them and add the deps myself)

**Comment by zkochan** [comment]
> how do you make the releases in the pnpm monorepo?

I do it currently manually.

> any pointers to suitable replacements for npm-registry-fetch and libnpmaccess would be appreciated.

for npm-registry-fetch you may use the pnpm alternative - [fetch-from-npm-registry](https://github.com/pnpm/pnpm/tree/master/packages/fetch-from-npm-registry). It currently handles only get requests but if you need other methods, they can probably be added.

I don't know alternatives for libnpmaccess. I'd recommend to fork it.

**Comment by evocateur** [comment]
> I do it currently manually.

Heh, ok, hopefully `lerna` will be able to help you there soon.

> for npm-registry-fetch you may use the pnpm alternative

I do use `PUT`, and also there are a few other deps like `pacote` that also use `npm-registry-fetch`. I'll probably try to nudge Kat, see if she can spare the bandwidth to accept a PR + publish a fix...

Thanks for all your hard work on pnpm!

**Comment by matthewharwood** [comment]
@evocateur I'm super interested in this... I run a pretty large mono repo and i'm hopeful that pnpm speeds up the bootstrap of packages!  Could you update this thread when ever you make any progress?

**Comment by evocateur** [comment]
I ended up forking n-r-f and the handful of libs with implicit deps. [pnpm-noodling](https://github.com/lerna/lerna/tree/pnpm-noodling) is looking pretty good. I absolutely _love_ `pnpm m outdated` and updating things is a breeze. There are a few hitches I've noticed in `pnpm recursive`, and I need to open issues once I've got an isolated repro...

**Comment by zkochan** [comment]
FYI, I fixed 3 very hard to reproduce bugs in `pnpm recursive install` in version `3.5.7`.

**Comment by evocateur** [comment]
@zkochan Yes! I have noticed some very nice improvements recently. I think https://github.com/pnpm/pnpm/issues/1504 is the only thing blocking me from switching to `pnpm` for all Zillow internal stuff.

**Comment by zkochan** [comment]
ok, I can prioritize that. I'll do it after the `workspace:` protocol (https://github.com/pnpm/pnpm/pull/1920)

**Comment by evocateur** [comment]
Fantastic, thanks!

**Comment by zkochan** [comment]
@evocateur the `pnpm outdated` command for scoped packages is now fixed in [v3.6.1](https://github.com/pnpm/pnpm/releases/tag/v3.6.1)

**Comment by evocateur** [comment]
@zkochan Thanks!

**Comment by vdtn359** [comment]
Are there any updates on this?

**Comment by georgyfarniev** [comment]
Hello. Any updates?

**Comment by baflo** [comment]
anything

**Comment by zkochan** [comment]
Seems like the WIP branch is called [pnpm-noodling](https://github.com/lerna/lerna/tree/pnpm-noodling)

**Comment by baflo** [comment]
There isn't much happening either :/

**Comment by allyraza** [comment]
why we don't move lerna and pnpm to npm instead of further dividing the community, I know its healthy to have a competition in the game but I think way too much competition is also not good yarn and npm is enough for now lets make npm great by merging back this functionality to npm

**Comment by georgyfarniev** [comment]
@allyraza I like your idea, and I'm pretty sure that authors of lerna and pnpm too, but I'm afraid that things are not that simple. Just merging this changes may break things due to how pnpm works.

**Comment by zkochan** [comment]
@evocateur let me know if I can help with this.

**Comment by evocateur** [comment]
I think I have it mostly ready? A lot of the struggles were with pnpm v3 and below, which required switching to forked packages with correct dependencies (already done and published). `pnpm-noodling` is mostly me dog-fooding the "latest" pnpm in the lerna monorepo itself. I may or may not end up merging that (it's a really nice dev experience, for sure!).

**Comment by kevinoneill** [comment]
How can I help validate and test pnpm support?

**Comment by stephenjang** [comment]
Is there any update?

**Comment by DesignByOnyx** [comment]
I too am glad to validate pnpm support. We are using it on a kind of large (and about to triple in size) monorepo which became too much for yarn's hoisting algorithm. Disabling hoisting didn't even work. PNPM just worked... flawlessly... and was 3x faster on installs. If there are any specific use cases you want me to try out, I'm glad to guinea pig.

**Comment by tomByrer** [comment]
> We are using it

@DesignByOnyx mind documenting how you got pnpm with Lerna please?

**Comment by DesignByOnyx** [comment]
Honestly, pnpm is capable of doing most of what lerna can do. Since my previous comment, we were able to replace all of our lerna commands with pnpm equivalents, and things just worked. I haven't tried publishing packages yet, so I can't speak for that part of the process - but I would bet that it's as smooth as everything else. Honestly my advice to anybody would be to switch to pnpm - and I say that with all due respect to the lerna project and its contributors. Seriously, lerna is amazing and has singlehandedly changed project/package management. I consider my advice to switch to pnpm an indicator that the lerna project has achieved the ultimate success.

Here are some major bullet points for our switch:
- pnpm is being used by the Microsoft Sharepoint team for multiple projects, and apparently by the Zillow team (see [comment](https://github.com/lerna/lerna/issues/1818#issuecomment-512976429) above), and I'm sure some other high-profile projects.
- pnpm has developed a large enough community to convince me it's safe
- I could not find any bad comments about pnpm despite my very deliberate attempts to do so. The only "bad" thing I found is that it's not as popular as yarn (meh).
- pnpm installs in roughly 3-4 seconds, while lerna + yarn takes about 16 - 20 seconds - so our CI pipelines are faster.
- pnpm revealed 2 accidental "implicit" dependencies we had in two different packages where a developer neglected to add the dependency directly to the package. Yarn's hoisting allows these kind of mistakes to go unnoticed.

**Comment by P4sca1** [comment]
@DesignByOnyx How do you handle released in your monorepo without Lerna?

**Comment by zkochan** [comment]
I am now trying to configure [changesets](https://github.com/atlassian/changesets) for publishing the pnpm monorepo. There are a few issues that needs fixing but the project seems decent, and I think their approach to versioning is scalable for projects that are used by big teams.

**Comment by DesignByOnyx** [comment]
I haven't really tried "conventional" publishing yet (eg. to npm) as our team manages artifacts in a custom way. But the pnpm docs on [filtering](https://pnpm.js.org/en/filtering) and [publishing](https://pnpm.js.org/en/cli/publish) makes me think anything is possible, especially `--filter [since]`.

**Comment by miripiruni** [comment]
Hi,

is there any progress here?

I research and compare tools for large frontend monorepo management (1500+ packages). For such large repo I want fast bootstrap. Since lerna works only with npm it's bootstrap is slow in compare with Rush+pnpm for example. But we already uses lerna and switch to Rush is relatively expansive.

**Comment by zkochan** [comment]
@miripiruni the switch to changesets+pnpm is easy: https://pnpm.js.org/en/using-changesets

If you need help, join the pnpm chatroom on discord or open issues in our repo.

**Comment by DmitryMasley** [comment]
I wonder, can we install deps using pnpm, but still use lerna scripts with `--since` flag. So kind of use both.

**Comment by zkochan** [comment]
pnpm has since as well https://pnpm.io/filtering#--filter-since

**Comment by DmitryMasley** [comment]
@zkochan it is not the same as `lerna`'s. `Lerna` detects changes since last tagged release instead of specific commit. We can workaround it, and write own logic to calculate last tagged commit, though. Also we could continue to use `lerna` with `conventional commits` support.

I wonder if there are any pitfalls here.

**Comment by EWhite613** [comment]
Some news from nwrl https://github.com/lerna/lerna/discussions/3140#discussioncomment-2973695

**Comment by DmitryMasley** [comment]
I managed to switch to `pnpm + lerna`. `pnpm` is used for bootstrap and dependency management and `lerna` is for publish. The only issue is after versions bump `pnpm` lock file also needs to be updated. I'm using a `version` script to do `pnpm install --lockfile-only --registry=my-registry && git add pnpm-lock.yaml`. Need  to add `registry` manually because `lerna` ignores `npmrc` in `version` script.

**Comment by ghiscoding** [comment]
You are correct in skipping lerna bootstrap when using pnpm, letting the package manager do the linking of all dependencies is the best way of dealing with this. For updating the pnpm lock file you can also use `postversion` since Lerna execute any post/pre scripts as part of its version [lifecycle scripts](https://github.com/lerna/lerna/tree/main/commands/version#lifecycle-scripts).

However, just to provide an alternative, you could use [Lerna-Lite](https://github.com/ghiscoding/lerna-lite) which I forked from Lerna a couple months before Lerna came back alive, I wish I'd knew before though, but anyway the only differences with Lerna vs Lerna-Lite is that I purposely didn't include a few commands (like `link` and `bootstrap`) and made a few optional, the commands are run the same way. I also added support for `workspace:` protocol that can be used with pnpm and latest yarn. I also dealt with the pnpm lock file as well recently, I simply added an optional flag (not enabled by default) to update the necessary lock file by leveraging the package manager itself (npm, pnpm or yarn), for example pnpm will execute `pnpm install --lockfile-only` at the end of the version processing cycle. I also keep PRs in sync with Lerna, this is now working so great with pnpm that I even switched the Lerna-Lite project itself to pnpm and also use `workspace:` protocol, everything is running smoothly after a few tweaks :)  

In short, wait for Lerna to support all pnpm missing pieces or give [Lerna-Lite](https://github.com/ghiscoding/lerna-lite) a try.
Cheers and happy coding

**Comment by ATQQ** [comment]
mark

**Comment by JamesHenry** [comment]
Hi Folks 👋 Closing this one as pnpm support was added a while ago, thanks all!
